rust/agama-server/src/cert.rs

```diff
@@ -1,84 +1,108 @@
-    let serial_number = {
-        let mut serial = BigNum::new()?;
-        serial.rand(159, MsbOption::MAYBE_ZERO, false)?;
-        serial.to_asn1_integer()?
-    };
-    builder.set_serial_number(&serial_number)?;
-    builder.set_subject_name(&x509_name)?;
-    builder.set_issuer_name(&x509_name)?;
-    builder.set_pubkey(&key)?;
-
-    let not_before = Asn1Time::days_from_now(0)?;
-    builder.set_not_before(&not_before)?;
-    let not_after = Asn1Time::days_from_now(365)?;
-    builder.set_not_after(&not_after)?;
-
-    builder.append_extension(BasicConstraints::new().critical().ca().build()?)?;
-
-    builder.append_extension(
-        SubjectAlternativeName::new()
-            // use the default Agama host name
-            // TODO: use the gethostname crate and use the current real hostname
-            .dns("agama")
-            // use the default name for the mDNS/Avahi
-            // TODO: check which name is actually used by mDNS, to avoid
-            // conflicts it might actually use something like agama-2.local
-            .dns("agama.local")
-            .build(&builder.x509v3_context(None, None))?,
-    )?;
-
-    let subject_key_identifier =
-        SubjectKeyIdentifier::new().build(&builder.x509v3_context(None, None))?;
-    builder.append_extension(subject_key_identifier)?;
-
-    builder.sign(&key, MessageDigest::sha256())?;
-    let cert = builder.build();
-
-    // for debugging you might dump the certificate to a file:
-    // use std::io::Write;
-    // let mut cert_file = std::fs::File::create("agama_cert.pem").unwrap();
-    // let mut key_file = std::fs::File::create("agama_key.pem").unwrap();
-    // cert_file.write_all(cert.to_pem().unwrap().as_ref()).unwrap();
-    // key_file.write_all(key.private_key_to_pem_pkcs8().unwrap().as_ref()).unwrap();
-
-    Ok((cert, key))
+const DEFAULT_CERT_DIR: &str = "/etc/agama.d/ssl";
+
+pub struct Certificate {
+    pub cert: X509,
+    pub key: PKey<Private>,
+}
+
+impl Certificate {
+    /// Writes cert, key to (for now well known) location(s)
+    pub fn write(&self) -> anyhow::Result<()> {
+        // check and create default dir if needed
+        if !Path::new(DEFAULT_CERT_DIR).is_dir() {
+            std::fs::create_dir_all(DEFAULT_CERT_DIR)?;
+        }
+
+        if let Ok(bytes) = self.cert.to_pem() {
+            fs::write(Path::new(DEFAULT_CERT_DIR).join("cert.pem"), bytes)?;
+        }
+        if let Ok(bytes) = self.key.private_key_to_pem_pkcs8() {
+            fs::write(Path::new(DEFAULT_CERT_DIR).join("key.pem"), bytes)?;
+        }
+
+        Ok(())
+    }
+
+    /// Reads cert from given path
+    pub fn read(cert: &Path, key: &Path) -> anyhow::Result<Self> {
+        let cert_bytes = std::fs::read(cert)?;
+        let key_bytes = std::fs::read(key)?;
+
+        let cert = X509::from_pem(&cert_bytes.as_slice());
+        let key = PKey::private_key_from_pem(&key_bytes.as_slice());
+
+        if let (Ok(c), Ok(k)) = (cert, key) {
```

Comment:
np: I would say using a `match` here is more idiomatic (https://rust-lang.github.io/rfcs/0160-if-let.html). But I have to admit that this one is a pedantic comment 😅